cmd/skopeo/completion/completions.go

```diff
@@ -0,0 +1,16 @@
+package completion
+
+import (
+	"github.com/containers/image/v5/transports"
+	"github.com/spf13/cobra"
+)
+
+// AutocompleteSupportedTransports list all supported transports with the colon suffix.
+func AutocompleteSupportedTransports(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
```

Comment:
For extra peace of mind (I’d rather not make an unexpected Go API stability promise), it would be nice not to make this available as a function that could be imported by third parties.

Probably simplest would be to move this to `cmd/skopeo.autocomplete…`. Alternatively, `cmd/skopeo/internal.Autocomplete…` or `cmd/skopeo/internal/completion.Autocomplete`. I don’t care at all which one is chosen.

Reply:
Is this actually a concern? Is there anywhere stated that skopeo code can be used as library and that it will keep the interfaces stable?
Also that function signature is already defined by the cobra library (since we have to set that function to the ValidArgsFunction) so it cannot be changed until upstream changes it.

Reply:
It’s a tiny concern, noting that https://github.com/containers/skopeo/blob/main/cmd/skopeo/inspect/output.go was explicitly defined as an external API, and it is side by side with this new subpackage. (Compare `go doc ./cmd/skopeo` (nothing) and `go doc ./cmd/skopeo/inspect`.)

Hypothetically, assuming we manage to make this completion code much smarter to also autocomplete the image names (for some transports), I can easily imagine e.g. Podman wanting to share the code — and making it private would be a reminder to move it to c/common instead.

Ultimately I feel fairly comfortable with telling anyone “this is not a supported API, I’m sorry, go away”, but we can make that situation impossible right now, at a trivial cost, so I think we might just as well do that.

--

The function’s _signature_ (and name) is not a problem at all, I’m talking purely about the functions’ external visibility. E.g. move the existing file to `cmd/skopeo/completions.go`, globally `s/AutocompleteSupported/autocompleteSupported/g`.

Reply:
done